### PR TITLE
TEST: update unit tests, and apply necessary fixes to library code, to make unit tests comply with mypy

### PR DESCRIPTION
**Pull request by j-ittner**

### CODE REVIEW

File: src/pytools/api/_api.py [hunk 1]
```diff
@@ -323,33 +400,58 @@ def validate_element_types(
...
+    else:
+        return type_
+
+
+def _raise_type_mismatch(
+    *,
+    name: Optional[str],
+    expected_type: Union[type, Tuple[type, ...]],
+    mismatched_type: type,
+    is_single: bool,
```
Comment (mtsokol) [inline src/pytools/api/_api.py:437]: I don't fully get why we need `is_single`. By "X expects instance of A" and "X expects instances of B" you indicate that a collection of such instances needs to be passed?

Comment (mtsokol) [reply src/pytools/api/_api.py:437]: Otherwise I don't see why such differentiation should be kept.

Comment (j-ittner) [reply src/pytools/api/_api.py:435]: @mtsokol good catch! It's a remnant of refactoring, and completely redundant.

Comment (j-ittner) [reply src/pytools/api/_api.py:437]: When called from `validate_type`, we are validating the type of a single object.

When called from `validate_element_types`, we are validating the type of multiple objects.

Arg `is_single` adjusts the wording of the error message depending on this difference.

Comment (mtsokol) [reply src/pytools/api/_api.py:437]: Then all clear to me!

**Review by mtsokol — APPROVED** [review]

**Pull request merged**